File: Problem-4/Solution.java
class Solution {
    public double findMedianSortedArrays(int[] nums1, int[] nums2) {
        if (nums1.length > nums2.length) {
            return findMedianSortedArrays(nums2, nums1);
        }

        int n = nums1.length, m = nums2.length;
        int left = 0, right = nums1.length;
        int totalNum = (n + m + 1) / 2;
        int i = 0, j = 0;
        while (left < right) {
            i = left + (right - left + 1) / 2;
            j = totalNum - i;

            if (nums1[i - 1] > nums2[j]) {
                right = i - 1;
            } else {
                left = i;
            }
        }

        i = left + (right - left + 1) / 2;
        j = totalNum - i;

        int nums1LeftMax = (i == 0) ? Integer.MIN_VALUE : nums1[i - 1];
        int nums1RightMin = (i == n) ? Integer.MAX_VALUE : nums1[i];
        int nums2LeftMax = (j == 0) ? Integer.MIN_VALUE : nums2[j - 1];
        int nums2RightMin = (j == m) ? Integer.MAX_VALUE : nums2[j];

        if ((n + m) % 2 == 1) {
            return Math.max(nums1LeftMax, nums2LeftMax);
        }
        return (double) (Math.max(nums1LeftMax, nums2LeftMax) + Math.min(nums1RightMin, nums2RightMin)) / 2;
    }
}

File: Problem-4/Solution.go
func max(a, b int) int {
	if a > b {
		return a
	}
	return b
}

func min(a, b int) int {
	if a < b {
		return a
	}
	return b
}

func findMedianSortedArrays(nums1 []int, nums2 []int) float64 {
	if (len(nums1) > len(nums2)) {
		return findMedianSortedArrays(nums2, nums1)
	}

	left, right := 0, len(nums1)
	totalNum := (len(nums1) + len(nums2) + 1) / 2
	var i, j int
	for (left < right) {
		i = left + (right - left + 1) / 2
		j = totalNum - i

		if (nums1[i - 1] > nums2[j]) {
			right = i - 1
		} else {
			left = i
		}
	}

	i = left + (right - left + 1) / 2
	j = totalNum - i

	var nums1LeftMax, nums1RightMin, nums2LeftMax, nums2RightMin int
	if (i == 0) {
		nums1LeftMax = math.MinInt64
	} else {
		nums1LeftMax = nums1[i - 1]
	}
	if (i == len(nums1)) {
		nums1RightMin = math.MaxInt64
	} else {
		nums1RightMin = nums1[i]
	}
	if (j == 0) {
		nums2LeftMax = math.MinInt64
	} else {
		nums2LeftMax = nums2[j - 1]
	}
	if (j == len(nums2)) {
		nums2RightMin = math.MaxInt64
	} else {
		nums2RightMin = nums2[j]
	}

	if ((len(nums1) + len(nums2)) % 2 == 1) {
		return float64(max(nums1LeftMax, nums2LeftMax)) 
	}
	return float64(max(nums1LeftMax, nums2LeftMax) + min(nums1RightMin, nums2RightMin)) / 2 
}